src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+                            System.Threading.Thread.MemoryBarrier();
+                            tmp.SecurityServiceDispatcher = this;
+                            this.sessionServerSettings = tmp;
+                        }
+                    }
+                }
+                return this.sessionServerSettings;
+            }
+        }
+
+        bool SupportsDuplex
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsDuplex;
+            }
+        }
+
+        bool SupportsRequestReply
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public bool SendUnsecuredFaults
+        {
+            get
+            {
+                return this.sendUnsecuredFaults;
+            }
+            set
+            {
+                this.sendUnsecuredFaults = value;
+            }
+        }
+
+        public IChannel OuterChannel
+        {
+            get { return this.outerChannel; }
+            set { this.outerChannel = value; }
+        }
+
+        IList<Type> IServiceDispatcher.SupportedChannelTypes => throw new NotImplementedException();
+
+        public ServiceHostBase Host =>  InnerServiceDispatcher.Host;
+
+        //private 
+
+        // This method should only be called at Open time, since it looks up the identity based on the 
+        // thread token
+        //void ComputeEndpointIdentity()
+        //{
+        //    EndpointIdentity result = null;
+        //    if (this.State == CommunicationState.Opened)
+        //    {
+        //        if (this.SecurityProtocolFactory != null)
+        //        {
+        //            result = this.SecurityProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //        else if (this.SessionServerSettings != null && this.SessionServerSettings.SessionProtocolFactory != null)
+        //        {
+        //            result = this.SessionServerSettings.SessionProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //    }
+        //    if (result == null)
+        //    {
+        //        result = base.GetProperty<EndpointIdentity>();
+        //    }
+        //    this.identity = result;
+        //}
+
+
+
+        void ThrowIfProtocolFactoryNotSet()
+        {
+            if (this.securityProtocolFactory == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecurityProtocolFactoryShouldBeSetBeforeThisOperation)));
+            }
+        }
+        public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync(IChannel outerChannel)
+        {
+            this.outerChannel = outerChannel;
+            IChannel securityChannel = GetInnerChannel(outerChannel);
```

Comment:
For Net.Tcp you can have this method being called twice for two different clients at the same time. This class represents the service as a whole. You can't store outerChannel as an instance variable as otherwise you will only be able to support one client at a time.

Reply:
done.. but let's review